beacon_node/beacon_chain/src/canonical_head.rs

```diff
@@ -0,0 +1,1307 @@
+    /// consider that these values represent the "canonical head" of the beacon chain.
+    pub cached_head: CanonicalHeadRwLock<CachedHead<T::EthSpec>>,
+    /// A lock used to prevent concurrent runs of `BeaconChain::recompute_head`.
+    ///
+    /// This lock **should not be made public**, it should only be used inside this module.
+    recompute_head_lock: Mutex<()>,
+}
+
+impl<T: BeaconChainTypes> CanonicalHead<T> {
+    /// Instantiate `Self`.
+    pub fn new(
+        fork_choice: BeaconForkChoice<T>,
+        snapshot: Arc<BeaconSnapshot<T::EthSpec>>,
+    ) -> Self {
+        let fork_choice_view = fork_choice.cached_fork_choice_view();
+        let forkchoice_update_params = fork_choice.get_forkchoice_update_parameters();
+        let cached_head = CachedHead {
+            snapshot,
+            justified_checkpoint: fork_choice_view.justified_checkpoint,
+            finalized_checkpoint: fork_choice_view.finalized_checkpoint,
+            head_hash: forkchoice_update_params.head_hash,
+            finalized_hash: forkchoice_update_params.finalized_hash,
+        };
+
+        Self {
+            fork_choice: CanonicalHeadRwLock::new(fork_choice),
+            cached_head: CanonicalHeadRwLock::new(cached_head),
+            recompute_head_lock: Mutex::new(()),
+        }
+    }
+
+    /// Load a persisted version of `BeaconForkChoice` from the `store` and restore `self` to that
+    /// state.
+    ///
+    /// This is useful if some database corruption is expected and we wish to go back to our last
+    /// save-point.
+    pub(crate) fn restore_from_store(
+        &self,
+        // We don't actually need this value, however it's always present when we call this function
+        // and it needs to be dropped to prevent a dead-lock. Requiring it to be passed here is
+        // defensive programming.
+        mut fork_choice_write_lock: RwLockWriteGuard<BeaconForkChoice<T>>,
+        store: &BeaconStore<T>,
+        spec: &ChainSpec,
+    ) -> Result<(), Error> {
+        let fork_choice = <BeaconChain<T>>::load_fork_choice(store.clone(), spec)?
+            .ok_or(Error::MissingPersistedForkChoice)?;
+        let fork_choice_view = fork_choice.cached_fork_choice_view();
+        let beacon_block_root = fork_choice_view.head_block_root;
+        let beacon_block = store
+            .get_full_block(&beacon_block_root)?
+            .ok_or(Error::MissingBeaconBlock(beacon_block_root))?;
+        let beacon_state_root = beacon_block.state_root();
+        let beacon_state = store
+            .get_state(&beacon_state_root, Some(beacon_block.slot()))?
+            .ok_or(Error::MissingBeaconState(beacon_state_root))?;
+
+        let snapshot = BeaconSnapshot {
+            beacon_block_root,
+            beacon_block: Arc::new(beacon_block),
+            beacon_state,
+        };
+
+        let fork_choice_view = fork_choice.cached_fork_choice_view();
```

Comment:
I don't think you meant to shadow your earlier version of `fork_choice_view` that you retrieved above right?